Versioning policy (#4085)
single-cell-data/TileDB-SOMA: dev_docs/POLICIES.md

# TileDB-SOMA Development Policies

## Version policy

TileDB-SOMA uses a loose variation of [semantic versioning](https://semver.org/) to organize and govern API compatibility, deprecations and version numbers.

TileDB-SOMA also uses lifecycle tags to indicate the maturity of an interface. All public API should have a lifecycle tag in docstrings. These tags are patterned after the RStudio lifecycle stage model. Tags are:

- ``experimental``: Under active development and may undergo significant and breaking changes.
- ``maturing``: Under active development but the interface and behavior have stabilized and are unlikely to change significantly but breaking changes are still possible.
- ``stable``: The interface is considered stable and breaking changes will be avoided where possible. Breaking changes that cannot be avoided will be accompanied by a major version bump.

> - ``deprecated``: The API is no longer recommended for use and may be removed in a future release.

Reviewer: ```suggestion
- ``deprecated``: The API is no longer recommended for use and will be removed in a future release.
```
The numpy documentation had a good point about not deprecating things that aren't going to be removed.

If no tag is present, the state is ``experimental``.

> [!NOTE]
> Prior to version 1.18, TileDB-SOMA had no explicit version policy. Versioning was most commonly `1.MAJOR.MINOR`,
> where breaking changes would occur in `1.X` releases, and any other change could occur on `1.X.Y` versions.

> > Starting with `1.18.0`, versioning will be aligned with the SemVer `MAJOR.MINOR.PATCH` schema.

Reviewer: I'm fairly certain will have a breaking change in the next release. Let's go ahead and assuming it will be 2.0 that we start using SemVer.

- Warnings of future compatibility changes, which are not "deprecate and remove", may be introduced in a `major` or `minor` release. Future compatibility warnings must be included in the appropriate change log.
- A `patch` release may not introduce deprecations or future incompatibility change warnings.

Experimental features may be modified in a `major` or `minor` release without prior warning. However, where possible, advance warning of changes will be provided.

## Warning period

Breaking API changes will provide notice of at least two `minor` releases or six months, whichever is greater, before effecting the change.

## Implementing Deprecation

### Python deprecations

Upon deprecation of Python functionality:

- a `DeprecationWarning` must be issued by the deprecated functionality and a unit test provided to confirm warning issuance.
- affected lifecyle tags must be set to "deprecated", typically in the docstring and other user-visible documentation.

> - docstrings must clearly indicate the scope of deprecation (e.g, function, a particular set of parameters, etc) and must include actionable guidance on migration.

Reviewer 2: ```suggestion
- docstrings must clearly indicate the scope of deprecation (e.g., function, a particular set of parameters, etc) and must include actionable guidance on migration.
```

- HISTORY.md must clearly indicate the deprecation and any guidance on migration.

For other breaking changes which are not "deprecate and remove":

- a `FutureWarning` must be issued by the deprecated functionality and a unit test provided to confirm warning issuance.
- docstrings must clearly indicate the scope of the breaking change (e.g, function, a particular set of parameters, etc) and must include actionable guidance on migration
- HISTORY.md must clearly indicate the breaking change and any guidance on migration.

Where entire classes, methods or functions are deprecated:

- utilize `typing_extensions.deprecated` decorator to provide both static and runtime warnings.
- update unit tests to ensure that the warning is issued (e.g., utilize `pytest.deprecated_call()`).
- ensure that `stacklevel` is set so that the warning comes from the invoking (caller's) code.

### R deprecations

Upon deprecation of R functionality, methods, or functions:

- a deprecation warning generated by `lifecycle::deprecate_warn()` must be issued by the deprecated functionality, and a unit test `lifecycle::expect_deprecated()` provided to confirm warning issuance.

> - if any parameters are being deprecated, they must be given a default value of `lifecycle::deprecated()`

Reviewer 2: ```suggestion
- if any parameters are being deprecated, they must be given a default value of `lifecycle::deprecated()`.
```

> - documentation must clearly indicate the scope of deprecation (e.g, function, a particular set of parameters, etc) and must include actionable guidance on migration.

Reviewer 2: ```suggestion
- documentation must clearly indicate the scope of deprecation (e.g., function, a particular set of parameters, etc) and must include actionable guidance on migration.
```

- `NEWS.md` must clearly indicate the deprecation and any guidance on migration.

> R functionality additionaly goes through a defunction process begining two `minor` releases after the deprecation

Reviewer 2: ```suggestion
R functionality additionally goes through a defunction process beginning two `minor` releases after the deprecation.
```

I think we should spell out what becoming defunct means in practice (will the function throw an error and stop working, or?).

For alignment what about the "or three months, whichever is greater"? If two minor releases happen in less than three months, it would be defunct before the three-month mark.

Maybe something like:

> Deprecated R functionality will become defunct
after two `minor` releases or 3 months, whichever is greater.

- a defunct error generated by `lifecycle::deprecate_soft()` must be issued, and a unit test `lifecycle::expect_defunct()` provided to confirm error issuance.
- documentation must clearly indicate the scope of defunction (e.g, function, a particular set of parameters, etc) and must include actionable guidance on migration.
- `NEWS.md` must clearly indicate the defunction and any guidance on migration.

At next major release after the defunction, the affected R functionality must be removed

For other breaking changes which are not "deprecate and remove":

> - a must be issued by the deprecated functionality and a unit test provided to confirm warning issuance.

Reviewer 2: Missing a word between *a* and *must*
